bridge-sdk/connectors/fast-bridge/src/fast_bridge.rs

```diff
@@ -0,0 +1,355 @@
+                recipient,
+                nonce,
+                amount,
+                unlock_recipient,
+                valid_till_block_height,
+            )
+            .value(amount);
+
+        let tx = transfer_call.send().await?;
+
+        tracing::info!(
+            tx_hash = format!("{:?}", tx.tx_hash()),
+            "Completed fast bridge transfer"
+        );
+
+        Ok(tx.tx_hash())
+    }
+
+    /// Unlocks tokens on Near following a successful transfer completion on Ethereum.
+    #[tracing::instrument(skip_all, name = "LP UNLOCK")]
+    pub async fn lp_unlock(&self, tx_hash: TxHash, log_index: u64) -> Result<CryptoHash> {
+        let eth_endpoint = self.eth_endpoint()?;
+        let near_endpoint = self.near_endpoint()?;
+
+        let proof = eth_proof::get_proof_for_event(tx_hash, log_index, eth_endpoint).await?;
+
+        let serialized_proof = serde_json::to_string(&proof).unwrap();
+        let args = format!(r#"{{"proof":{serialized_proof}}}"#)
+            .to_string()
+            .into_bytes();
+
+        tracing::debug!("Retrieved Ethereum proof");
+
+        let tx_hash = near_rpc_client::change(
+            near_endpoint,
+            self.near_signer()?,
+            self.fast_bridge_account_id()?.to_string(),
+            "lp_unlock".to_string(),
+            args,
+            120_000_000_000_000,
+            0,
+        )
+        .await?;
+
+        tracing::info!(
+            tx_hash = format!("{:?}", tx_hash),
+            "Sent lp unlock transaction"
+        );
+
+        Ok(tx_hash)
+    }
+
+    /// Withdraw tokens from the fast bridge contract.
+    #[tracing::instrument(skip_all, name = "WITHDRAW")]
+    pub async fn withdraw(
+        &self,
+        token_id: AccountId,
+        amount: Option<U128>,
+        recipient_id: Option<AccountId>,
+        msg: Option<String>,
+    ) -> Result<CryptoHash> {
+        let near_endpoint = self.near_endpoint()?;
+
+        let mut json = format!(r#"{{"token_id": "{token_id}""#);
+        if let Some(recipient_id) = recipient_id {
+            json.push_str(&format!(r#","recipient_id": "{recipient_id}""#));
+        }
+        if let Some(amount) = amount {
+            json.push_str(&format!(r#","amount": "{amount}""#));
+        }
+        if let Some(msg) = msg {
+            json.push_str(&format!(r#","msg": "{msg}""#));
+        }
+        json.push_str("}");
+
+        let args = json.to_string().into_bytes();
+
+        let tx_hash = near_rpc_client::change(
+            near_endpoint,
+            self.near_signer()?,
+            self.fast_bridge_account_id()?.to_string(),
+            "withdraw".to_string(),
+            args,
+            20_000_000_000_000,
+            0,
+        )
+        .await?;
+
+        tracing::info!(
+            tx_hash = format!("{:?}", tx_hash),
+            "Sent withdraw transaction"
+        );
+
+        Ok(tx_hash)
+    }
+
+    fn near_signer(&self) -> Result<near_crypto::InMemorySigner> {
```

Comment:
It seems to me that this code will be the same not only in fast-bridge; maybe it can be extracted to a common place.

Reply:
It's on my mind but since it's not a lot of code I'm copy-pasting for now